### ISSUE TITLE
Fast (non-normalized) inner product search?

### ISSUE THREAD

**Issue by jschmitz28** [searchivarius/nmslib#191]
Hello,

I've recently discovered nmslib and am very impressed!

Looking through the code and manual, I'm trying to figure out whether there is support for a fast/approximate non-normalized maximum inner product search. I've found this issue which appears to add support for normalized dot product: https://github.com/searchivarius/nmslib/issues/110

And I've also found an implementation of the exact distance function I'm interested in searching on: https://github.com/searchivarius/nmslib/blob/master/similarity_search/src/distcomp_scalar.cc#L274

But I can't find it in the table of implemented spaces. Is this supported, or would it be difficult to add?

Thanks!

**Comment by searchivarius** [comment]
Hi @jschmitz28 you need an inner-product search for dense vectors, right? Currently we support only sparse vectors, but supporting the dense inner-product search is easy to add.

**Comment by jschmitz28** [comment]
Thanks for the quick response! Yes, I'd be interested in dense vector support for inner product search.

**Comment by searchivarius** [comment]
I added **negdotprod** to develop. It is also possible to reduce inner product search to cosine similarity search, but this requires data transformation. The benefit is that HNSW has an especially optimized implementation for cosine.

**Comment by searchivarius** [comment]
Actually it seems that I forgot the sign https://github.com/searchivarius/nmslib/commit/3849cfa3b1e7c4414f52bfec1c05f40d1524be0e
I will have to look at this more carefully to make sure it works as intended.

**Comment by jschmitz28** [comment]
Awesome! Without knowing the code well it looks good to me - I wrote a basic unit test (d9665ba) which does the same sort of correctness verification as is done in test_lpnorm.cc for euclidean, and it looks like it works as expected (validated through bunit).

Regarding the reduction to cosine similarity, do you mean by first normalizing the vectors? The data I'm dealing with for this is the output of an ML pipeline for which we've been having some trouble achieving comparable performance when applying normalization.

**Comment by searchivarius** [comment]
@jschmitz28 thank you for checking. Would you submit a pull request? I had some doubts, b/c unnormlized inner product seems to be a harder distance than the normalized one. Although this wasn't the case for sparse vectors, it does make sense.

The reduction/normalization I talk about is invertible (which can be important for downstream components) and it produces the same set of neighbors. This ingenious yet simple transformation is described here http://proceedings.mlr.press/v37/neyshabur15.html
 
Basically you normalize each element by dividing its elements by the largest l2 in the dataset. You normalize queries by dividing elements by the query norm. Then you add an additional dimension. For the query it is zero. For a data point, it is equal to sqrt(1-squared norm of the data point).

**Comment by jschmitz28** [comment]
Added: https://github.com/searchivarius/nmslib/pull/192

Thanks for the link! That's super useful.

**Comment by searchivarius** [comment]
cool, glad this was useful.

**Comment by x10000year** [comment]
@searchivarius Does this method of additional dimension reduce the precision of query result? Especially when there are some outlier element that has extremely large l2 norm?

**Comment by searchivarius** [comment]
@x10000year thank you for the interest in our library, what do you mean exactly by the method? We have search methods/algorithms and the so-called spaces, which are data+distance.

**Comment by x10000year** [comment]
@searchivarius 

I mean the method you mentioned for process non-normalized maximum inner product search:
"Basically you normalize each element by dividing its elements by the largest l2 in the dataset. You normalize queries by dividing elements by the query norm. Then you add an additional dimension. For the query it is zero. For a data point, it is equal to sqrt(1-squared norm of the data point)."

If one outlier element has extremely large l2 norm, while most other elements have small l2 norm, does this method lead to very skewed data distribution? I wonder whether the performance of nmslib/hnsw is still good in this situation.

**Comment by searchivarius** [comment]
@x10000year NMSLIB doesn't use this method. I also don't know whether such a reduction affects performance. I have plans to check this, but I haven't done yet. In general, if you have outliers with the large l2 norm, you k-NN may become poorly defined, because outliers with large norms can often be huge attractors, i.e., they may be nearest neighbors for a majority of data set points. Needless to say this wouldn't be much useful.

**Comment by jattenberg** [comment]
How would I utilize this with the python API?

**Comment by searchivarius** [comment]
Hi @jattenberg what exactly do you need?

**Comment by jattenberg** [comment]
maybe NVM? generating recommendations. I think the correct py instantiation for an index using max inner product search would be: `index = nmslib.init(method='hnsw', space='negdotprod')`. Is this correct?

**Comment by searchivarius** [comment]
ok, please model after this notebook (but with space='negdotprod'):
https://github.com/nmslib/nmslib/blob/master/python_bindings/notebooks/search_vector_dense_nonoptim.ipynb
unfortunately, currently data needs to be reloaded.
